repair/row_level.cc

```diff
@@ -322,14 +329,77 @@ mutation_reader repair_reader::make_reader(
+                sstring filename;
+            };
+
+            std::vector<foreign_ptr<lw_shared_ptr<utils::chunked_vector<sstables::sstable_files_snapshot>>>> ptrs(smp::count);
+            auto sstable_filenames = co_await db.map_reduce0(
+                [token_range = _token_range, &ptrs, table_id = _schema->id()] (auto& db) -> future<std::vector<sst_and_shard>> {
+                    std::vector<sst_and_shard> res;
+                    auto& t = db.find_column_family(table_id);
+                    auto sstables = co_await t.take_storage_snapshot(token_range);
+                    for (auto& sst : sstables) {
+                        co_await coroutine::maybe_yield();
+                        rlogger.debug("Get sst filename={}", sst.sst->get_filename());
+                        res.push_back(sst_and_shard{this_shard_id(), sst.sst->get_filename()});
+                    }
+                    ptrs[this_shard_id()] = make_foreign(make_lw_shared<utils::chunked_vector<sstables::sstable_files_snapshot>>(std::move(sstables)));
+                    co_return res;
+                },
+                std::vector<sst_and_shard>{},
+                [] (std::vector<sst_and_shard> a, std::vector<sst_and_shard> b) { std::move(b.begin(), b.end(), std::back_inserter(a)); return a;
+            });
+            _sstable_snapshots = std::move(ptrs);
+
+            auto& t = db.local().find_column_family(_schema);
+            auto erm = t.get_effective_replication_map();
+            auto& sstm = t.get_sstables_manager();
+            _sstable_set = make_lw_shared<sstables::sstable_set>(sstables::make_partitioned_sstable_set(_schema, false));
+            size_t nr_local_sstables = 0;
+            size_t nr_remote_sstables = 0;
+            // Load sst from this shard into _sstable_set
+            auto& sstables_on_this_shard = _sstable_snapshots[this_shard_id()];
+            if (sstables_on_this_shard) {
+                for (auto& sst : *sstables_on_this_shard) {
+                    auto insert = _sstable_set->insert(sst.sst);
+                    rlogger.debug("Add local sst filename={} insert={}", sst.sst->get_filename(), insert);
+                    ++nr_local_sstables;
+
+                }
+            }
+            // Load sstables from other shard into _sstable_set
+            for (auto& _sst_and_shard : sstable_filenames) {
+                if (_sst_and_shard.shard == this_shard_id()) {
+                    continue;
+                }
+                auto& filename = _sst_and_shard.filename;
+                auto data_path = std::filesystem::path(filename);
+                auto desc = sstables::parse_path(data_path, _schema->ks_name(), _schema->cf_name());
+                auto sst = sstm.make_sstable(t.schema(), t.dir(), t.get_storage_options(),
+                        desc.generation, sstables::sstable_state::normal, desc.version, desc.format);
+                co_await sst->load(erm->get_sharder(*_schema));
```

Comment:
I don't know if this is "legal" or not.
@raphaelsc can we load an sstable on multiple shards? Any expected fallout?

Reply:
it's fine to load a sstable on a remote shard, as long as only the owner shard deletes it.

I am actually worried about memory usage here. with vnodes, shard ranges aren't contiguous, so sstables from different shards overlap one another. with ics / stcs, there will be multiple sstables overlapping with each token, on every shard. let's say we have 96 shards, each shard having 10 sstables (even with ics, there will be 1 sstable in each run overlapping with any given token). the memory usage can go 10×96×128k(buffer_size)×4(read_ahead)=~500M. with m5d.24xlarge, that's 12% of memory for a shard. I haven't reviewed the patch, but it seems we now get rid of evictable reader, so we no longer suffer with thrashing, but the increased memory usage can translate into increased latency since it requires evicting data from the cache, right?

Reply:
If we read from all shards we still need 500M of memory but each shard only uses 500M / 96, right? Any way we could reduce the memory usage? We could reduce the read_ahead. How about the buffer_size, can we reduce it too? We need a very conservative reader for this.

Reply:
I think this memory consumption concern is indeed very real and it can easily cause OOM if we are not careful.

Additional memory consumption will come from just loading the sstables. We have several components that we keep in memory at all times. We can have a special flag for the open function to only load the bare minimum (e.g. not load the bloom-filter). The summary can also be large and there is no way to cut it out because we need it to read from the index, to read the ranges from the sstables.